src/response.rs

```diff
@@ -507,18 +562,44 @@ impl Response {
 
         let url = unit.url.clone();
 
-        Ok(Response {
+        // Prefetch data if it's small enough to be buffered in one go.
+        let peeked_len = stream.fill_buf()?.len();
+        let is_fully_buffered = Some(peeked_len) == length;
+
+        let resp = Response {
             url,
             status_line,
             index,
             status,
             headers,
             unit: Box::new(unit),
-            stream: Mutex::new(Box::new(stream.into())),
+            stream: Mutex::new(ResponseStream::Empty),
```

Comment:
It feels awkward to construct the Response with `Empty` (an invalid state!) and then replace it with one or the other thing later. Can we construct the ResponseStream first, and then build Response with it directly?

Reply:
Not easily. The only way to get the stream back to the pool, is to read the buffered response via the full reader. I looked at refactoring the `PoolReturnRead` logic, but it's very involved – depends on `Stream`.

This is why `make_read` does not take `self`

Reply:
Ah, that makes sense.

Rather than refactoring the PoolReturnRead logic, what about bypassing / duplicating it? Once we've concluded that we're reading by Content-Length, the logic is simple: we take N bytes from the BufReader, reset the stream's deadlines, and call `agent.state.pool.add(&PoolKey::new(url, agent.config.proxy.clone(), stream)`.

Reply:
Groan. More pooling code? :)

Do you think it will be simpler than what we got here?

Reply:
Heh, yeah, more pooling code is a downside. I think it could possibly be simpler, in part because it lets us play to the strengths of BufReader: namely that you get access to a slice of bytes, then inform BufReader "hey, I took N of those bytes," avoiding any `.read()` call.

But I'm okay with the create-and-modify.